src/main/java/frc/robot/subsystems/vision/VisionSubsystem.java

```diff
@@ -1,175 +1,235 @@
-    currentlyUsedLimelightResults = LimelightHelpers.getLatestResults(currentlyUsedLimelight);
-
-    // Turns the limelight LEDs on if they can't see an april tag
-    if (!canSeeAprilTags()) {
-      LimelightHelpers.setLEDMode_ForceOn(VisionConstants.FRONT_LEFT_LIMELIGHT_NAME);
-      LimelightHelpers.setLEDMode_ForceOn(VisionConstants.FRONT_RIGHT_LIMELIGHT_NAME);
-    } else {
-      LimelightHelpers.setLEDMode_ForceOff(VisionConstants.FRONT_LEFT_LIMELIGHT_NAME);
-      LimelightHelpers.setLEDMode_ForceOff(VisionConstants.FRONT_RIGHT_LIMELIGHT_NAME);
+  /**
+   * This makes a thread which updates the pose estimate for the specified limelight. Because we
+   * have 3 limelights, if we are trying to parse the JSON dump for each limelight one after another
+   * every scheduler loop, we will have loop overruns, so we opted to have them run in parallel like
+   * this.
+   *
+   * @param limelightNumber the limelight number
+   */
+  public void visionThread(int limelightNumber) {
+    poseLock.lock();
+    try {
+      new Thread(
+              () -> {
+                double last_TX = 0;
+                double last_TY = 0;
+
+                while (true) {
+                  double current_TX = LimelightHelpers.getTX(getLimelightName(limelightNumber));
+                  double current_TY = LimelightHelpers.getTY(getLimelightName(limelightNumber));
+
+                  // This checks if the limelight reading is new. The reasoning being that if the TX
+                  // and TY
+                  // are EXACTLY the same, it hasn't updated yet with a new reading. We are doing it
+                  // this way,
+                  // because to get the timestamp of the reading, you need to parse the JSON dump
+                  // which can be
+                  // very demanding whereas this only has to get the Network Table entries for TX
+                  // and TY.
+                  if (last_TX != current_TX || last_TY != current_TY) {
+                    updateLimelightPoseEstimate(limelightNumber);
+                    // This is to keep track of the last valid pose calculated by the limelights
+                    // it is used when the driver resets the robot odometry to the limelight
+                    // calculated position
+                    if (canSeeAprilTags(limelightNumber)) {
+                      lastSeenPose = getMegatag1Pose(limelightNumber);
+                    }
+                  } else {
+                    limelightEstimates[limelightNumber] = new PoseEstimate();
+                  }
+
+                  last_TX = current_TX;
+                  last_TY = current_TY;
+                  try {
+                    Thread.sleep(20);
+                  } catch (Exception e) {
+                    e.printStackTrace();
+                  }
+                }
+              })
+          .start();
+    } catch (Exception e) {
+      e.printStackTrace();
+    } finally {
+      poseLock.unlock();
```

Comment:
how does the locking work? All of the stuff is still happening in a while loop, so does it ever reach here?

Weren't we talking about making it so that every vision subsystem execute, it makes a thread for each limelight that checks for updates in pose exactly one time, and then shuts down?

I might have misunderstood stuff tho so idk

Reply:
plz check my fixes for ur second point @JacksonElia